feat: Add tekton-bundle support

lcarva opened the pull request:
## Changes

This adds support for a new package manager, [Tekton Bundles](https://tekton.dev/docs/pipelines/pipelines/#tekton-bundles).

## Context

Tekton Bundles are a mechanism to use Pipeline and Task definitions in [Tekton](https://tekton.dev) that are stored in an OCI image. Context about this feature can be found in the initial [proposal](https://github.com/tektoncd/community/blob/main/teps/0005-tekton-oci-bundles.md).

The scope of this package manager is only Tekton Bundles. Support for updating Pipeline and Task definitions that are provided via other mechanisms (e.g. inline) are purposely excluded. If support for those is desired, I suggest a separate package manger as they are distinct enough to require a completely separate approach to extracting dependencies.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

rarkins commented:
I think the code is lacking any reference to tekton.dev so far? We should link to it, e.g. from the Readme to introduce what it is.

It's OK for a manager to have limited support (e.g. support docker images and not git, etc) but the Readme should also clearly document that as a limitation so that they are considered a *feature* to be added and not a *bug* to report.

I don't prefer the idea of a separate package manager in future though if the unsupported dependency types are part of the same Tekton concept and files - logically that should be the same manager.

lcarva commented:
@rarkins, thank you for the feedback. Pushed a commit that, I believe, addresses your concerns:

```
- Rename manager from `tekon-bundle` to `tekton`. The dependency type is
still `tekton-bundle`. This leaves room for extending the manager to
support different dependency types in the future, e.g. git references.

- Update docs to introduce Tekton and provide additional context, as well
as clarify which features are supported.
```

viceice commented on lib/modules/manager/tekton/index.ts:
```diff
@@ -0,0 +1,120 @@
+import { DockerDatasource } from '../../datasource/docker';
+import type { PackageDependency, PackageFile } from '../types';
+
+export const language = ProgrammingLanguage.Docker;
```
remove this, as it will interfere with future non docker deps

viceice commented on lib/modules/manager/tekton/index.ts:
```diff
@@ -0,0 +1,120 @@
+export const language = ProgrammingLanguage.Docker;
+
+export const defaultConfig = {
+  // Tekton uses YAML files to define its kubernetes resources. These
+  // don't have a specific file name pattern. To avoid issues with
+  // unrelated YAML files, the match pattern is left empty by default.
+  // Users must provide this value based on their own Tekton usage.
+  // For example, to match all the YAML files in a repository, add
+  // the following to renovate.json in the corresponding git repo:
+  //  {
+  //    "tekton": {
+  //      "fileMatch": ['\\.yaml$', '\\.yml$']
+  //    }
+  //  }
```
remove this, it's already in readme

viceice commented on lib/modules/manager/tekton/index.ts:
```diff
@@ -0,0 +1,120 @@
+  return deps;
+}
+
+function createDep(imageRef: string): PackageDependency | null {
```
reuse `getDep` from dockerfile manager

viceice commented on lib/modules/manager/tekton/index.ts:
```diff
@@ -0,0 +1,120 @@
+  const deps: PackageDependency[] = [];
+  let docs: any[];
+  try {
+    docs = loadAll(content) as any[];
```
please create proper typescript interfaces

lcarva replied:
Unfortunately, I don't have a full set of all the possible types the yaml document could represent. There is also no common set of attributes between those documents that satisfy this.

The required attributes could be found in the PipelineRun and TaskRun resources. These could be defined in their own yaml document, or could be nested under the TriggerTemplate resource. I can create 3 or 4 different Typescript interfaces to cover these.

There's also the case where they could be defined across multiple yaml documents, and files, when using something like [kustomize](https://github.com/kubernetes-sigs/kustomize), in which case we'll lose some flexibility if using typescript interfaces (I think).

Initially, this seemed like a good case to **not** use a typescript interface, but I'm happy to change it if that's more inline with the project. 

Given this information, would you still like me to change the code to use typescript interfaces?

viceice replied:
you don't need to define all possible interfaces and properties, just the required ones so we've more type safety. 

i don't think we loose any flexibility. it's just more work.

lcarva replied:
Very well. Updated the code to do so.

viceice commented on lib/modules/manager/tekton/index.ts:
```diff
@@ -0,0 +1,120 @@
+
+export const supportedDatasources = [DockerDatasource.id];
+
+export function extractPackageFile(
```
move to a `./extract.ts` file

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,97 @@
+): PackageFile | null {
+  logger.trace('tekton.extractPackageFile()');
+  const deps: PackageDependency[] = [];
+  let docs: any[];
```
```suggestion
  let docs: TektonResource[];
```

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,97 @@
+  const dep = getDep(imageRef);
+  // If a tag is not found, assume the lowest possible version. This will
+  // ensure the version update is successful, and properly pin the digest.
+  dep.currentValue = dep.currentValue ?? '0.0';
```
if it's a docker image, it doesn't need a `currentValue`, docker datasource will use `latest` docker tag instead.

lcarva replied:
Good call! Removed this line and was able to verify the expected behavior was still met. Thanks.

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,97 @@
+  const deps: PackageDependency[] = [];
+
+  // Handle TaskRun resource
+  addDep(doc?.spec?.taskRef, deps);
```
doc should never undefined or null here, so please all following
```suggestion
  addDep(doc.spec?.taskRef, deps);
```

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,97 @@
+    );
+    return null;
+  }
+  for (const doc of docs) {
```
filter docs here, if there can be null / undefiend docs. I'm not sure if this can really happen
```suggestion
  for (const doc of docs.filter(is.truthy)) {
```

I'm also not sure if `loadAll` can return `null`, please check other managers, which use this function

lcarva replied:
I moved this to `getDeps` so I could also completely remove the usage of `doc?` within it.

viceice commented on lib/modules/manager/tekton/__fixtures__/multi-doc.yaml:
```diff
@@ -0,0 +1,146 @@
+---
```
i think you can remove some of the fixtures docs without loosing coverage. 

all of them are docker images, so you should only test the doc structure and not every permutation of a docker image reference

lcarva replied:
Sure thing. Removed some cases.

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,97 @@
+    );
+    return null;
+  }
+  for (const doc of docs) {
```
```suggestion
  if (!is.array(docs)) {
    return null;
  }

  for (const doc of docs) {
```
i think `loadAll` returns null for an empty string. please validate with a test.

lcarva replied:
Added unit test to verify behavior when file is empty (empty string). `loadAll` does not return null. I believe the function signature also confirms this:
```typescript
export function loadAll(str: string, iterator?: null, opts?: LoadOptions): unknown[];
```

The suggested change would introduce unreachable code thus failing to meet the required 100% unit test coverage.

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,97 @@
+function addDep(ref: TektonBundle, deps: PackageDependency[]): void {
+  let imageRef = '';
+  // Find a bundle reference from the Bundle resolver
+  if (ref?.resolver === 'bundles') {
```
please add an early null check for ref, like you did for doc in the other function

lcarva replied:
Much better, yes. Done.

viceice reviewed (commented):
please update your branch from main with a merge commit. your jest snapshot uses an outdated format.

lcarva commented:
> please update your branch from main with a merge commit. your jest snapshot uses an outdated format.

Done!

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,100 @@
+  if (!is.string(imageRef) || is.emptyString(imageRef)) {
+    // Fallback to older style bundle reference
+    imageRef = ref.bundle;
+    if (!is.string(imageRef) || is.emptyString(imageRef)) {
```
there is a `is.notEmptyAndNotWhitespaceString` function to combine both conditions. 😉

lcarva replied:
I don't see that one. I see `nonEmptyStringAndNotWhitespace` and `emptyStringOrWhitespace` which are slightly different and don't guard against the case where `imageRef` is not a string at all which I think is something important to guard against.

I'm surprised that imageRef in some cases could be something other than a string given that the types have been defined. This sort of seems like an odd behavior in TypeScript. For this document:

```
kind: Pipeline    
spec:             
  tasks:          
  - taskRef:      
      bundle: true
```

`typeof imageRef` returns `'boolean'`. However, `is.boolean(imageRef)` returns `false`. I'm digging more into this.

lcarva replied:
Hmm it seems that this is a limitation where the type enforcement is not done at runtime. I'll change the code to assume it's a string, and let the dependency have a `skipReason: invalid-value`. This is likely the most correct way to approach this as at that point, there's an invalid Tekton resource defined.

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,100 @@
+      }
+    }
+  }
+  if (!is.string(imageRef) || is.emptyString(imageRef)) {
```
see comment below

rarkins reviewed (changes requested):
Please update the Readme formatting:
- No manual line wrapping
- Use one sentence per line

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,101 @@
+
+  // Handle list of TektonResources
+  const items = doc.items ?? [];
+  if (items.length > 0) {
```
remove this condition and inline items, like all other loops

viceice commented on lib/modules/manager/tekton/index.spec.ts:
```diff
@@ -0,0 +1,32 @@
+import { Fixtures } from '../../../../test/fixtures';
```
rename file to extract.spec.ts

viceice commented on lib/modules/manager/tekton/extract.spec.ts:
```diff
@@ -0,0 +1,32 @@
+    it('extracts deps from a file', () => {
+      expect(
+        extractPackageFile(Fixtures.get('multi-doc.yaml'), 'test-file.yaml')
+      ).toMatchSnapshot();
```
add assertion for deps length

lcarva replied:
The snapshot performs an exact match, which covers this and more. Wouldn't that be a redundant check?

viceice replied:
no, snapshots tend to be silently updated with wrong changes

lcarva replied:
Added length check.

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  // Handle list of TektonResources
+  const items = doc.items ?? [];
```
inline items

viceice commented on lib/modules/manager/tekton/__snapshots__/extract.spec.ts.snap:
```diff
@@ -0,0 +1,133 @@
+      "depType": "tekton-bundle",
+      "replaceString": "gcr.io/tekton-releases/catalog/upstream/pipeline-run-resolver",
+    },
+    {
+      "depType": "tekton-bundle",
+      "skipReason": "invalid-value",
+    },
+    {
+      "depType": "tekton-bundle",
+      "skipReason": "invalid-value",
+    },
```
should they've been filtered because they have no depName?

lcarva replied:
No. This is a side effect of implementing the suggested [change](https://github.com/renovatebot/renovate/pull/17289#discussion_r955313352), see discussion there.

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,98 @@
+    }
+  }
+
+  if (is.emptyStringOrWhitespace(imageRef)) {
```
```suggestion
  if (!imageRef || is.emptyStringOrWhitespace(imageRef)) {
```
so we need the additional check again

lcarva replied:
This will not catch the case where imageRef is not a string, but a truthy value. If we want a proper check, we should revert back to using `is.string` since TypeScript does not perform type checking at runtime. We can also create a comprehensive type guard function. This seems sort of heavy handed, especially because I have not seen this approach in other places of the code base.

I quite like the current approach as it highlights that an attribute was not processed because it has an unexpected type, which I believe is what `invalid-value` is meant to cover.

viceice replied:
ok, then we should somplify this here to:
```suggestion
  if (is.nullOrUndefined(imageRef)) {
```

is `imageRef` is isn't defined, fallback to old style.

lcarva replied:
That doesn't handle when the string is empty, which results in a dependency like this being detected:
```typescript
{
  "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
  "currentDigest": undefined,
  "currentValue": undefined,
  "datasource": "docker",
  "depName": "",
  "depType": "tekton-bundle",
  "replaceString": "",
},
```

I'll be honest that I don't quite understand why we want to simplify this further. As mentioned in the previous comment, using `is.emptyStringOrWhitespace` seems to provide the most correct behavior, and it's already a single call.

lcarva replied:
@viceice, any additional feedback here? :pray:

viceice replied:
it should not fallback if `imageRef` is defined, even it's not a string value.

although the current returned invalid values aren't helpful, as you don't know where they com from. so we should filter any non string and empty or whitespace strings.

lcarva replied:
ok, I think I understand your concern now. thank you for clarifying. Updated the code.

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,98 @@
+  if (is.emptyStringOrWhitespace(imageRef)) {
+    // Fallback to older style bundle reference
+    imageRef = ref.bundle;
+    if (is.emptyStringOrWhitespace(imageRef)) {
```
```suggestion
    if (!imageRef || is.emptyStringOrWhitespace(imageRef)) {
```

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,98 @@
+  if (is.falsy(ref)) {
+    return;
+  }
+  let imageRef = '';
```
```suggestion
  let imageRef : string | undefined;
```

allow undefined

viceice reviewed (commented):
see comment

viceice commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,98 @@
+  if (is.nullOrUndefined(imageRef)) {
+    // Fallback to older style bundle reference
+    imageRef = ref.bundle;
+    if (is.nullOrUndefined(imageRef)) {
+      return;
+    }
+  }
```
```suggestion
  }
```

Docker `getDep` will handle it.

- #17644

rarkins commented on lib/modules/manager/tekton/extract.ts:
```diff
@@ -0,0 +1,95 @@
+    imageRef = ref.bundle;
+  }
+
+  const dep = getDep(imageRef);
```
`imageRef` could be `undefined` here if `ref.bundle` is `undefined`. Perhaps need to add  a check using `is` that it's a non empty string?

viceice replied:
nope, will now be handled in `getDep`
